### PR TITLE
add(aggregate parsing functions)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

add aggegration functions to metrics parser to provide values for the whole flow cell.
Also gets a total number of undetermined reads for that flow cell.

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 1]
```diff
@@ -175,3 +175,62 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+            if demux_metric.sample_internal_id == UNDETERMINED:
+                aggregate_undetermined_read_pairs += demux_metric.read_pair_count
+        return self.calculate_total_reads_for_metrics(
+            read_pair_count=aggregate_undetermined_read_pairs
+        )
+
+    def get_aggregate_percent_q30_for_metrics(self) -> float:
+        """Calculate the aggregate percent Q30 for the demux metrics excluding indexchecks."""
+        aggregate_q30_bases: int = 0
+        non_index_count: int = 0
```
Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:205]: so these files also contain "indexchecks" under the "Sample_Project" column. I figured we do not want these.

Comment (ChrOertlin) [reply cg/services/bcl_convert_metrics_service/parser.py:205]: nvm we don't use the Sample_project column any longer.. we have so need to change this

Comment (ChrOertlin) [reply cg/services/bcl_convert_metrics_service/parser.py]: also it seems we do not report on index checks any longer. So just summing is good enough

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 2]
```diff
@@ -175,3 +175,54 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+
+            aggregate_yield += quality_metric.yield_
+        return aggregate_yield
+
+    def get_aggregate_yield_q30_for_metrics(self) -> int:
+        """Calculate the aggregate yield Q30 for the demux metrics excluding indexchecks."""
+        aggregate_yield_q30: int = 0
+        for quality_metric in self.quality_metrics:
+            aggregate_yield_q30 += quality_metric.yield_q30
+        return aggregate_yield_q30
```
Comment (diitaz93) [inline cg/services/bcl_convert_metrics_service/parser.py:228]: This snippet is called several times in this class, could we turn it into a method? something like:
```python
def get_aggregate_for_attribute(self, obj_list: list, attr_name: str) -> int:
        """Calculate the aggregate sum of a specified attribute for a given list of objects."""
        aggregate_value: int = 0
        for obj in obj_list:
            aggregate_value += getattr(obj, attr_name)
        return aggregate_value
```

File: tests/fixtures/services/bcl_convert_metrics_service/230622_A00621_0864_AHY7FFDRX2/Unaligned/Reports/Adapter_Metrics.csv
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply tests/fixtures/services/bcl_convert_metrics_service/230622_A00621_0864_AHY7FFDRX2/Unaligned/Reports/Adapter_Metrics.csv:1]: These don't have undetermined reads in them. Having the sample project column there or not does not affect anything.

File: cg/services/bcl_convert_metrics_service/models.py [hunk 3]
```diff
@@ -13,7 +13,7 @@ class SequencingQualityMetrics(BaseModel):
     )
     q30_bases_percent: float = Field(..., alias=QualityMetricsColumnNames.Q30_BASES_PERCENT)
     yield_: int = Field(..., alias=QualityMetricsColumnNames.YIELD)
-    yield_q30: float = Field(..., alias=QualityMetricsColumnNames.YIELD_Q30)
+    yield_q30: int = Field(..., alias=QualityMetricsColumnNames.YIELD_Q30)
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/services/bcl_convert_metrics_service/models.py:16]: I saw that tis was fixed in another PR 👍

**Review by diitaz93 — APPROVED** [review]
Great 💯

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 4]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
         )
         return bool(metrics)
+
+    @classmethod
+    def calculate_total_reads_for_metrics(cls, read_pair_count: int) -> int:
+        """Scale to read pair number up to single reads."""
+        return read_pair_count * SCALE_TO_READ_PAIRS
+
+    def get_aggregate_total_reads_for_metrics(self):
+        """Return the aggregate reads for the whole demux metrics excluding indexchecks."""
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py]: I see no logic regarding index checks
```suggestion
        """Return the aggregate reads for the whole demux metrics."""
```

Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:177]:
```suggestion
        """Return the aggregate reads for the whole demux metrics."""
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 5]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        aggregate_read_pairs: int = self.get_aggregate_for_attribute(
+            metrics=self.demux_metrics, attr_name="read_pair_count"
+        )
+        return self.calculate_total_reads_for_metrics(read_pair_count=aggregate_read_pairs)
+
+    def get_aggregate_undetermined_reads_for_metrics(self) -> int:
+        """Calculate the total undetermined reads to the demux metrics excluding indexchecks."""
```
Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:184]:
```suggestion
        """Calculate the total undetermined reads to the demux metrics."""
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 6]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        aggregate_undetermined_read_pairs: int = 0
+        for demux_metric in self.demux_metrics:
+            if demux_metric.sample_internal_id == UNDETERMINED:
+                aggregate_undetermined_read_pairs += demux_metric.read_pair_count
+        return self.calculate_total_reads_for_metrics(
+            read_pair_count=aggregate_undetermined_read_pairs
+        )
+
+    def get_aggregate_percent_q30_for_metrics(self) -> float:
+        """Calculate the aggregate percent Q30 for the demux metrics excluding indexchecks."""
```
Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:194]:
```suggestion
        """Calculate the aggregate percent Q30 for the demux metrics."""
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 7]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        aggregate_q30_bases: int = self.get_aggregate_for_attribute(
+            metrics=self.quality_metrics, attr_name="q30_bases_percent"
+        )
+        return round(aggregate_q30_bases / len(self.quality_metrics), 2)
+
+    def get_aggregate_quality_score_for_metrics(self) -> float:
+        """Calculate the aggregate quality score for the demux metrics excluding indexchecks."""
```
Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:201]:
```suggestion
        """Calculate the aggregate quality score for the demux metrics."""
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 8]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        aggregate_quality_score: int = self.get_aggregate_for_attribute(
+            self.quality_metrics, "mean_quality_score_q30"
+        )
+        return round(aggregate_quality_score / len(self.quality_metrics), 2)
+
+    def get_aggregate_yield_for_metrics(self) -> int:
+        """Calculate the aggregate yield for the demux metrics excluding indexchecks."""
```
Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:208]:
```suggestion
        """Calculate the aggregate yield for the demux metrics."""
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 9]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        return self.get_aggregate_for_attribute(metrics=self.quality_metrics, attr_name="yield_")
+
+    def get_aggregate_yield_q30_for_metrics(self) -> int:
+        """Calculate the aggregate yield Q30 for the demux metrics excluding indexchecks."""
```
Comment (ChrOertlin) [inline cg/services/bcl_convert_metrics_service/parser.py:212]:
```suggestion
        """Calculate the aggregate yield Q30 for the demux metrics."""
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 10]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        return self.calculate_total_reads_for_metrics(
+            read_pair_count=aggregate_undetermined_read_pairs
+        )
+
+    def get_aggregate_percent_q30_for_metrics(self) -> float:
+        """Calculate the aggregate percent Q30 for the demux metrics."""
+        aggregate_q30_bases: int = self.get_aggregate_for_attribute(
+            metrics=self.quality_metrics, attr_name="q30_bases_percent"
+        )
+        return round(aggregate_q30_bases / len(self.quality_metrics), 2)
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py:198]: I think this is a problematic way to aggregate this metric. An aggregate percentage metric will only be the sum of it's part if all parts are the same size. We would need to normalise this somehow. Will come back to this

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 11]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+
+    def get_aggregate_quality_score_for_metrics(self) -> float:
+        """Calculate the aggregate quality score for the demux metrics."""
+        aggregate_quality_score: int = self.get_aggregate_for_attribute(
+            self.quality_metrics, "mean_quality_score_q30"
+        )
+        return round(aggregate_quality_score / len(self.quality_metrics), 2)
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py:205]: Same for this one

**Review by Vince-janv — CHANGES_REQUESTED** [review]
Some concerns on two of the aggregate metrics. Will look into it

File: cg/services/bcl_convert_metrics_service/parser.py (hunk 10 above)
Comment (ChrOertlin) [reply cg/services/bcl_convert_metrics_service/parser.py:198]: Yes the name is a bit misleading... this is indeed the mean_q30_percent not the aggregate.

File: cg/services/bcl_convert_metrics_service/parser.py (hunk 11 above)
Comment (ChrOertlin) [reply cg/services/bcl_convert_metrics_service/parser.py:205]: Same here. Will change it to mean.

**Comment by ChrOertlin** [comment]
# tests

Ensured `cg demultiplex finish flow-cell` runs without errors

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 12]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        """Calculate the total undetermined reads to the demux metrics."""
+        aggregate_undetermined_read_pairs: int = 0
+        for demux_metric in self.demux_metrics:
+            if demux_metric.sample_internal_id == UNDETERMINED:
+                aggregate_undetermined_read_pairs += demux_metric.read_pair_count
+        return self.calculate_total_reads_for_metrics(
+            read_pair_count=aggregate_undetermined_read_pairs
+        )
+
+    def get_mean_percent_q30_for_aggregated_metrics(self) -> float:
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py:193]: Can we rename the "aggregate" functions to "for_flow_cell"? I feel like in some of these there are so many terms that one looses track of what the function does.
```suggestion
    def get_percent_q30_for_flow_cell(self) -> float:
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 13]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        """Calculate the mean percent Q30 for the aggregated quality metrics."""
+        aggregate_yield: int = self.get_aggregate_yield_for_metrics()
+        aggregate_yield_q30: int = self.get_aggregate_yield_q30_for_metrics()
+        return round(aggregate_yield_q30 / aggregate_yield, 2)
+
+    def get_mean_quality_score_sum_for_aggregated_metrics(self) -> float:
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py:199]:
```suggestion
    def get_mean_quality_score_for_flow_cell(self) -> float:
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 14]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
...
+        return read_pair_count * SCALE_TO_READ_PAIRS
+
+    def get_aggregate_total_reads_for_metrics(self):
+        """Return the aggregate reads for the whole demux metrics."""
+        aggregate_read_pairs: int = self.get_aggregate_for_attribute(
+            metrics=self.demux_metrics, attr_name="read_pair_count"
+        )
+        return self.calculate_total_reads_for_metrics(read_pair_count=aggregate_read_pairs)
+
+    def get_aggregate_undetermined_reads_for_metrics(self) -> int:
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py:183]:
```suggestion
    def get_undetermined_reads_for_flow_cell(self) -> int:
```

File: cg/services/bcl_convert_metrics_service/parser.py [hunk 15]
```diff
@@ -175,3 +167,55 @@ def has_undetermined_reads_in_lane(self, lane: int) -> bool:
             metrics=self.quality_metrics, sample_internal_id=UNDETERMINED, lane=lane
         )
         return bool(metrics)
+
+    @classmethod
+    def calculate_total_reads_for_metrics(cls, read_pair_count: int) -> int:
+        """Scale to read pair number up to single reads."""
+        return read_pair_count * SCALE_TO_READ_PAIRS
+
+    def get_aggregate_total_reads_for_metrics(self):
```
Comment (Vince-janv) [inline cg/services/bcl_convert_metrics_service/parser.py:176]:
```suggestion
    def get_total_reads_for_flow_cell(self):
```

**Review by Vince-janv — APPROVED** [review]
Looks great ⭐

**Pull request merged**